src/main/scala/com/fulcrumgenomics/util/zip/Zipper.scala

```diff
@@ -0,0 +1,840 @@
+      logger.debug(" (U=" + unusedQueue.size() + ", C=" + toConsumeQueue.size() + ", W=" + toWriteQueue.size() +
+        ", R=" + reader.blocksBuffered + ")"
+      )
+      */
+    }
+
+    // throw the exception if one was raised in a thread
+    threadException.map { thr => throw thr }
+
+    // stop all runnables
+    require(reader.isStopped) // should be done, so need to stop it
+    consumers.foreach { consumer => consumer.stopIt() }
+    writer.stopIt()
+
+    // close the queues
+    unusedQueue.close()
+    toConsumeQueue.close()
+    toWriteQueue.close()
+
+    // join
+    threads.foreach { thread => thread.join() } // TODO: timeout?
+
+    // forces the writer to perform any last steps (ex. write the last empty GZIP block).
+    writer.finish()
+
+    // stop itself
+    stopIt()
+  }
+
+  /** Closes the input and output channels, and any other streams opened using them. */
+  def close(): Unit = {
+    reader.close()
+    writer.close()
+  }
+}
+
+/** Companion methods for input and output blocks of data. */
+private[zip] object Blocks {
+  /** Returns the approximate size in bytes for Blocks */
+  val sizeInBytes: Long = {
+    SequentialBlocks.sizeInBytes(true) + SequentialBlocks.sizeInBytes(false)
+  }
+
+  /** The default block length in bytes of uncompressed data. */
+  val DefaultUncompressedFixedBlockLength = DEFAULT_UNCOMPRESSED_BLOCK_SIZE
+
+  /** The maximum block length in bytes of compressed data (ignores the GZIP header). */
+  val MaxCompressedFixedBlockLengthNoHeader = MAX_COMPRESSED_BLOCK_SIZE - BLOCK_HEADER_LENGTH
+
+  /** The maximum block length in bytes of compressed data (with the GZIP header and footer). */
+  val MaxCompressedFixedBlockLength = MAX_COMPRESSED_BLOCK_SIZE - 2 // subtract two for the BGZF extra field
+
+  /** Companion definitions and methods for sequential blocks. */
+  private[zip] object SequentialBlocks {
+
+    /** The default # of blocks in a sequential block. */
+    private[zip] var DefaultCapacity: Int = 128
+
+    /** Returns the approximate size in bytes for Blocks. */
+    def sizeInBytes(compress: Boolean, capacity: Int = DefaultCapacity): Long = {
+      if (compress) MaxCompressedFixedBlockLength * capacity
+      else DefaultUncompressedFixedBlockLength * capacity
+    }
+  }
+
+  /**
+    * Stores multiple blocks to compress or decompress.  This facilitates more efficient reading and writing
+    * of the data, since more data can be read or written at any one time.  It also allows for multiple blocks to
+    * be compressed or decompressed by a single consumer.
+    *
+    * @param capacity the capacity of fixed-size blocks.
+    * @param compress true if the data will be compressed, false if it is to be decompressed.
+    */
+  private class SequentialBlocks private(capacity: Int, compress: Boolean) {
+    def this(compress: Boolean) = {
+      this(capacity=SequentialBlocks.DefaultCapacity, compress=compress)
```

Comment:
Why use a secondary constructor to do this instead of just making capacity defaulted in the primary constructor?